File: e-commerce/app/src/main/java/gparap/apps/e_commerce/adapters/CategoryAdapter.java
/*
 * Copyright 2022 gparap
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gparap.apps.e_commerce.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import gparap.apps.e_commerce.R;
import gparap.apps.e_commerce.data.CategoryResponseModel;
import gparap.apps.e_commerce.ui.ProductActivity;
import gparap.apps.e_commerce.utils.AppConstants;

public class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.CategoryViewHolder> {
    private ArrayList<CategoryResponseModel> categories = new ArrayList<>();
    private Context context;

    @SuppressWarnings("unused")
    public ArrayList<CategoryResponseModel> getCategories() {
        return categories;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setCategories(ArrayList<CategoryResponseModel> categories) {
        this.categories = categories;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //get context
        context = parent.getContext();

        //get layout inflater
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        //create item view
        View view = inflater.inflate(R.layout.cardview_category, parent, false);

        //create the view holder for this item view
        return new CategoryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryViewHolder holder, int position) {
        //create the image Uri
        String uriString = AppConstants.IMAGES_URL + categories.get(position).getUrl();
        Uri uri = Uri.parse(uriString);

        //set the category image
        Picasso.get().load(uri).into(holder.imageView);

        //show products of this category in another activity
        holder.imageView.setOnClickListener(view->{
            Intent intent = new Intent(context, ProductActivity.class);
            intent.putExtra(AppConstants.CATEGORY_ID, categories.get(position).getId());
            context.startActivity(intent);
        });
    }

    @Override
    public int getItemCount() {
        return categories.size();
    }

    static class CategoryViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageView;

        public CategoryViewHolder(@NonNull View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.image_view_category_main);
        }
    }
}

File: shopping-list/app/src/main/java/gparap/apps/shopping_list/adapters/CategoryAdapter.kt
/*
 * Copyright 2021 gparap
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gparap.apps.shopping_list.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import gparap.apps.shopping_list.R
import gparap.apps.shopping_list.data.CategoryModel

class CategoryAdapter(private val callback: CategoryAdapterCallback) :
    RecyclerView.Adapter<CategoryAdapter.CategoryViewHolder>() {
    var categories = ArrayList<CategoryModel>()
    lateinit var context: Context

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        context = parent.context
        val layoutInflater =
            parent.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = layoutInflater.inflate(R.layout.cardview_categories, parent, false)
        return CategoryViewHolder(view)
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        //display category info
        holder.categoryName.text = categories[position].name.toString()
        holder.itemsCount.text = context.resources.getString(R.string.text_items_count)
            .plus(categories[position].itemsCount.toString())

        //handle itemView clicks
        holder.itemView.setOnClickListener {
            callback.onCategoryViewClickListener(categories[position])
        }

        //handle edit button clicks
        holder.editCategory.setOnClickListener {
            callback.onEditCategoryButtonClickListener(categories[position])
        }

        //handle delete button clicks
        holder.deleteCategory.setOnClickListener {
            callback.onDeleteCategoryButtonViewClickListener(categories[position])
        }
    }

    override fun getItemCount(): Int {
        return categories.size
    }

    class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var categoryName = itemView.findViewById<TextView>(R.id.text_view_category)!!
        var itemsCount = itemView.findViewById<TextView>(R.id.text_view_category_items_count)!!
        val editCategory = itemView.findViewById<ImageView>(R.id.image_view_edit_category)!!
        val deleteCategory = itemView.findViewById<ImageView>(R.id.image_view_delete_category)!!
    }

    /** Callback for listening to RecyclerView's item clicks */
    interface CategoryAdapterCallback {
        fun onCategoryViewClickListener(category: CategoryModel)
        fun onEditCategoryButtonClickListener(category: CategoryModel)
        fun onDeleteCategoryButtonViewClickListener(category: CategoryModel)
    }
}